File: ml-benchmark/synduce/zipper/list_sum.f
Inductive List = nil Unit | cons {Int, List};
Inductive Zipper = zip {List, List};

concat = fix (
  \f: List -> List -> List. \xs: List. \ys: List.
  match xs with
    nil _ -> ys
  | cons {h, t} -> cons {h, f t ys}
  end
);

rev = fix (
  \f: List -> Compress List. \xs: List.
  match xs with
    nil _ -> nil unit
  | cons {h, t} -> concat (f t) (cons {h, nil unit})
  end
);

list_repr = fix (
  \f: List -> Compress List. \xs: List.
  match xs with
    nil _ -> nil unit
  | cons {h, t} -> cons {h, f t}
  end
);

sum = fix (
  \f: List -> Int. \xs: List.
  match xs with
    nil _ -> 0
  | cons {h, t} -> + h (f t)
  end
);

repr = \z: Zipper.
  match z with
    zip {l, r} -> concat (rev l) (list_repr r)
  end;

main = \z: Zipper. sum (repr z);

File: ml-benchmark/synduce/zipper/list_sum.ml
type list =
  | Nil of unit
  | Cons of int * list

type zipper =
  | Zip of list * list

let rec concat xs ys =
  match xs with
  | Nil _ -> ys
  | Cons (h, t) -> Cons (h, concat t ys)

let rec rev xs =
  match xs with
  | Nil _ -> Nil ()
  | Cons (h, t) -> concat (rev t) (Cons (h, Nil ()))

let rec list_repr xs =
  match xs with
  | Nil _ -> Nil ()
  | Cons (h, t) -> Cons (h, list_repr t)

let rec sum xs =
  match xs with
  | Nil _ -> 0
  | Cons (h, t) -> h + sum t

let repr z =
  match z with
  | Zip (l, r) -> concat (rev l) (list_repr r)

let program z = sum (repr z)